### ISSUE TITLE
Blocked read statement through Bolt Java Driver

### ISSUE THREAD

**Issue by vigorouscoding** [neo4j/neo4j#12033]
Neo4j version: 3.4.7
Operating system: Mac OS X 10.13.4 (and Ubuntu 18.04 LTS)
API/Driver: (Bolt Java Driver v1.6.2)

## Description

When querying a "big" result set (around 1000) nodes and then - in the same transaction - querying elements of this result set, Neo4j blocks the result completely and does not return at all when using the `.list()` function.

## Steps to reproduce

Run the following test class against a Neo4j 3.4.7 instance. The magic number of nodes seems to be 1002 (see the `for` loop). After each test I used the command `MATCH (n) DETACH DELETE n` to clear the database.

The output will only be the first node id `Node: <nodeId>`.

```` java
public class BlockTest {

	public static void main(final String[] args) {

		try (final Driver driver = GraphDatabase.driver("bolt://localhost:7687",
						AuthTokens.basic("neo4j", "admin"),
						Config.build().withEncryptionLevel(Config.EncryptionLevel.NONE).toConfig()
			)) {

			try (final Session session = driver.session()) {

				// create the test data
				try (final Transaction tx = session.beginTransaction()) {

					tx.run("CREATE (n:FirstType {id:\"abcde12345\" }) RETURN n");

					for (int i = 0; i < 1002; i++) {
						tx.run("MATCH (n:FirstType {id:\"abcde12345\" }) CREATE (n)-[:CONNECTION]->(x:SecondType {id:\"secondType_" + i + "\" }) RETURN x");
					}

					tx.success();
				}

				Integer count = 0;

				// query the test data
				try (final Transaction tx = session.beginTransaction()) {

					final List<Record> x = tx.run("MATCH (n:FirstType {id:\"abcde12345\" })-[:CONNECTION]->(x:SecondType) RETURN x").list();

					for (final Record r : x) {

						count++;

						final NodeValue n = (NodeValue)r.get("x");
						System.out.println("Node: " + n.asNode().id());

						// this is where the block occurs
						tx.run("MATCH (x:SecondType)<-[r:CONNECTION]-(y) WHERE ID(x) = " + n.asNode().id() + " RETURN y").list();
					}

					tx.success();
				}

				System.out.println("Records: " + count);
			}
		}
	}
}
````

## Expected behavior

The query should return the requested elements.

## Actual behavior

The thread blocks - see the following `jstack` output.

````
"main" #1 prio=5 os_prio=31 tid=0x00007ff6f0800000 nid=0x1603 waiting on condition [0x0000700003714000]
   java.lang.Thread.State: WAITING (parking)
        at sun.misc.Unsafe.park(Native Method)
        - parking to wait for  <0x0000000717b06e08> (a java.util.concurrent.CompletableFuture$Signaller)
        at java.util.concurrent.locks.LockSupport.park(LockSupport.java:175)
        at java.util.concurrent.CompletableFuture$Signaller.block(CompletableFuture.java:1693)
        at java.util.concurrent.ForkJoinPool.managedBlock(ForkJoinPool.java:3323)
        at java.util.concurrent.CompletableFuture.waitingGet(CompletableFuture.java:1729)
        at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1895)
        at org.neo4j.driver.internal.util.Futures.blockingGet(Futures.java:109)
        at org.neo4j.driver.internal.InternalStatementResult.blockingGet(InternalStatementResult.java:123)
        at org.neo4j.driver.internal.InternalStatementResult.list(InternalStatementResult.java:94)
        at com.mycompany.neoblocked.NeoBlock.main(NeoBlock.java:53)
````

**Comment by lutovich** [comment]
Hi @vigorouscoding,

Thanks for reporting this problem!
I believe this is an issue in the Java driver related to network-level backpressure applied when receiving incoming records. However, blocking does not happen with this exact example on my machine.

Could you please run this example with `-DrecordBufferHighWatermark=10000` system property? And then change 1002 to 10002?

This system property increases the record buffer size, which has the default value of 1000. Driver stops reading from the network when buffer size is greater than this value.

Looking forward to hearing from you.

**Comment by vigorouscoding** [comment]
Hi @lutovich,

thanks for getting back to me - I just re-tested this and found that while cleaning up my test code I messed up. I'm really sorry about that!

The original code that fails looks like the following - the only difference is that I did not use `list()` to fetch the results. But now, reading more of the code, it seems this might be user error (`Results are valid until the next statement is run or until the end of the current transaction, whichever comes first. To keep a result around while further statements are run, or to use a result outside the scope of the current transaction, see list().`). I'm really sorry that I provided the wrong source and even more so if this is my mistake :(


```
	try (final Driver driver = GraphDatabase.driver("bolt://localhost:7687",
					AuthTokens.basic("neo4j", "admin"),
					Config.build().withEncryptionLevel(Config.EncryptionLevel.NONE).toConfig()
		)) {

		try (final Session session = driver.session()) {

			// create the test data
			try (final Transaction tx = session.beginTransaction()) {

				tx.run("CREATE (n:FirstType {id:\"abcde12345\" }) RETURN n");

				for (int i = 0; i < 1002; i++) {
					tx.run("MATCH (n:FirstType {id:\"abcde12345\" }) CREATE (n)-[:CONNECTION]->(x:SecondType {id:\"secondType_" + i + "\" }) RETURN x");
				}

				tx.success();
			}

			Integer count = 0;

			// query the test data
			try (final Transaction tx = session.beginTransaction()) {

				final StatementResult x = tx.run("MATCH (n:FirstType {id:\"abcde12345\" })-[:CONNECTION]->(x:SecondType) RETURN x");

				while (x.hasNext()) {

					final Record r = x.next();

					count++;

					final NodeValue n = (NodeValue)r.get("x");
					System.out.println("Node: " + n.asNode().id());

					// this is where the block occurs
					tx.run("MATCH (x:SecondType)<-[r:CONNECTION]-(y) WHERE ID(x) = " + n.asNode().id() + " RETURN y").list();
				}

				tx.success();
			}

			System.out.println("Records: " + count);
		}
	}
```

Just for the record - I tried the `-DrecordBufferHighWatermark=10000` parameter and it did not change the blocking behaviour, but this might not matter if it was user error.

**Comment by lutovich** [comment]
Hi @vigorouscoding,

Such nested queries within a single transaction should be supported and it is a bug in the Java driver. Mentioned PR should fix the problem.

Could you please try out the fix? It is located in [this branch](https://github.com/lutovich/neo4j-java-driver/tree/1.6-tx-with-nested-queries) which can be pulled in as a Maven dependency using [JitPack](https://jitpack.io/#lutovich/neo4j-java-driver), just to avoid building it locally.

Thanks again for reporting this problem!

**Comment by vigorouscoding** [comment]
Hi @lutovich,

this looks very good - I simply added the following configuration to my `pom.xml`, cleaned, re-compiled and it worked every time. I even went up to 10000 nodes to give it something proper to chew on - took a moment longer but worked just as well.
Only thing is that I had to use the short commit ID in the dependency - I couldn't get it to work otherwise.

```
		<repository>
			<id>jitpack.io</id>
			<url>https://jitpack.io</url>
		</repository>
```

and

```
	<dependency>
	    <groupId>com.github.lutovich</groupId>
	    <artifactId>neo4j-java-driver</artifactId>
	    <version>a1504c1</version>
	</dependency>
```

Thanks a lot - and I especially like the really long commit message so I was able to get a deeper understanding what happened internally!
